APID 2.15.0: Changes for the GitHub integration (#139)
mulesoft/docs-design-center: modules/ROOT/pages/design-storage-github.adoc

= Storing Projects in GitHub
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

:imagesdir: ../assets/images

> By default, projects that you create in API Designer are saved in Design Center. However, you can save a project as a repository in GitHub.

Reviewer: Should we make clear in the previous text that although you have these two options, the default is Design Center?

The following diagram depicts the steps to follow to configure an Anypoint Platform business organization (APBO) to use a GitHub organization as a repository for API Designer projects. It also depicts which users are allowed to access the GitHub organization:

.This diagram shows the steps for enabling and using the support for saving projects in GitHub.
image:apid-flow-for-GitHub.png[The steps for enabling and using the support for GitHub]

As depicted on the far left of the diagram, one person with _Design Center Developer_ permission in your APBO installs the MuleSoft application in GitHub organization X. This person installs the application as part of creating the first API Designer project in the APBO to use GitHub.

. Click *Create* to start creating a project in the text editor.
+
You can create either an API-specification project that you want to work on in the text editor, or you can create an API-fragment project. The visual editor stores projects in Design Center.
. Install the MuleSoft application in a GitHub organization.
+

> Click the *Install from GitHub* option to start the process for installing the MuleSoft application in a GitHub organization. As part of the process, you log into GitHub and select the organization to install the application in.

Reviewer: log in to:
```suggestion
Click the *Install from GitHub* option to start the process for installing the MuleSoft application in a GitHub organization. As part of the process, you log in to GitHub and select the organization to install the application in.
```

+
After you finish the installation, you return to the dialog for creating a project.
. In the *Save location* field, select your GitHub ID for the organization that you installed the MuleSoft application in.

> . Finish filling in the fields of the dialog and click the *Create* button.

Reviewer: ```suggestion
. Complete the fields of the dialog and click the *Create* button.
```

After the one-time installation, all developers in the APBO that have the _Design Center Developer_ permission and access to GitHub organization X can also create API Designer projects as repositories in that GitHub organization. They follow the same steps, but without following the steps of the installation procedure.

> Developers who do not have access to GitHub organization X, of course, cannot create projects as repositories in that organization, and they cannot edit projects that are saved there. However, developers who do have access to GitHub organization X but are not in the APBO can edit the repositories and create repositories to import into API Designer as projects, even though they cannot create projects through Design Center.

Reviewer: ```suggestion
Developers who do not have access to GitHub organization X cannot create projects as repositories in that organization, and they cannot edit projects that are saved there. However, developers who do have access to GitHub organization X but are not in the APBO can edit the repositories and create repositories to import into API Designer as projects, even though they cannot create projects through Design Center.
```

== Open a Project that is Saved as a Repository in GitHub

Projects saved in GitHub are listed in Design Center together with projects that are saved in Design Center. They are indicated by the GitHub icon at the end of a project's name.

== Import a Project from GitHub

You can import an API-specification project or an API-fragment project that is stored as a repository in GitHub, whether you created the project in the text editor in API Designer or in an editor outside of API Designer.

. Open Design Center from the homepage of Anypoint Platform.
. On the *Projects* page in Design Center, click *Create* and then click *Create API Specification*.
. Click *Import* near the top of the dialog.
+
The dialog lists the projects that you can import from the GitHub organizations that your user ID belongs to.
. Select the project that you want to import.
+

> If the project does not include an exchange.json file that was generated by the text editor in API Designer, a dialog is opened to ask you to specify which file is the root file of the project.

Reviewer: ```suggestion
If the project does not include an `exchange.json` file that was generated by the text editor in API Designer, a dialog is opened to ask you to specify which file is the root file of the project.
```

The project is imported and opened in the text editor.

== Save an Existing API Designer Project as a Repository in GitHub

If you have one or more API Designer projects that are saved in Design Center and you would like to save them instead in GitHub, follow these steps to migrate each one:

. Export a copy of a project from Design Center.
. Create in GitHub a repo for the copy of the project, but give the repo a name that is different from the name of the original project.
. Extract the content of the downloaded ZIP file.

> . In the exchange.json file, edit the `name` field to match the name of the repo.

Reviewer: ```suggestion
. In the `exchange.json` file, edit the `name` field to match the name of the repo.
```

. Add the extracted content to the repository.
. Import the project from GitHub.
. Delete the original project.

== Working with Branches

> Every time that you open a project that is saved in GitHub, API Designer retrieves the current content of the main branch. Every time that you switch to a different branch, API Designer retrieves the current content of that branch.

Reviewer: I suggest not switching between "main branch" and "default branch."
Try to stick with only one term for a thing.

Reviewer: Also, in both sentences, you could replace "Every time that you..." with "When you...".

> The branch that you are currently working in is displayed at the top of the central pane of the text editor. Click the name of the branch to open a menu to carry out any of the following tasks:

Reviewer: Try to replace even common idioms:
```suggestion
The branch that you are currently working in is displayed at the top of the central pane of the text editor. Click the name of the branch to open a menu to perform any of the following tasks:
```

* Switch to a different branch by clicking the name of that branch.
+

> Every time you open a project that is stored in GitHub as a repository, the content of the default branch appears in the text editor. So, if you want to work in a non-default branch, you must switch to that branch.

Reviewer: Do we need to repeat?

Reviewer: ```suggestion
Every time you open a project that is stored in GitHub as a repository, the content of the default branch appears in the text editor. So, if you want to work in a nondefault branch, you must switch to that branch.
```

+

> Before switching, you can view the difference between that branch and the current one.

Reviewer: ```suggestion
Before switching, you can view the difference between the branch you want to switch to and the current one.
```

* Create a new branch.
+
After you create a branch, the text editor switches to that branch.
* Open a pull request.
+
This option behaves exactly as the *Create Pull Request* option in GitHub Desktop. It opens a page in GitHub where you can specify the name of your pull request and add a comment before clicking *Create*.
* View the current branch on GitHub.
* Delete a branch.

== Managing Conflicting Changes

> If another user commits changes to the branch in which you are working, the text editor displays a dialog to inform you that is will discard your most recent changes (within the last 15 seconds or so) and incorporate the commits from the other user. Therefore, only one person at a time should work in a branch.

Reviewer: ...and incorporate the changes
or
...and incorporate the commit

> == Removing from Design Center an API Designer Project that is Saved in GitHub

Reviewer: Please use the Chicago tab of the online capitalization tool:
```suggestion
== Removing from Design Center an API Designer Project That Is Saved in GitHub
```
